File: data/readable_format/cpp/file673.cpp
#include <iostream>
#include <cmath>
#include <vector>

class RandomNumberGenerator {
public:
    RandomNumberGenerator() : RDM(31), RM1(0), RM2(0), IA1(1279), IC1(351762), M1(1664557),
                              IA2(2011), IC2(221592), M2(1048583), IA3(15091), IC3(6171), M3(29201),
                              IX1(0), IX2(0), IX3(0) {}

    void DRNSET(int ISEED) {
        IX1 = ISEED % M1;
        IX1 = (IA1 * IX1 + IC1) % M1;
        IX2 = IX1 % M2;
        IX1 = (IA1 * IX1 + IC1) % M1;
        IX3 = IX1 % M3;
        RM1 = 1.0 / static_cast<double>(M1);
        RM2 = 1.0 / static_cast<double>(M2);
        for (int J = 0; J < 31; ++J) {
            IX1 = (IA1 * IX1 + IC1) % M1;
            IX2 = (IA2 * IX2 + IC2) % M2;
            RDM[J] = (static_cast<double>(IX1) + static_cast<double>(IX2) * RM2) * RM1;
        }
    }

    double DRN(int& ISEED) {
        IX1 = (IA1 * IX1 + IC1) % M1;
        IX2 = (IA2 * IX2 + IC2) % M2;
        IX3 = (IA3 * IX3 + IC3) % M3;
        int J = 1 + (31 * IX3) / M3;
        double result = RDM[J - 1]; // Adjust for 0-based indexing
        RDM[J - 1] = (static_cast<double>(IX1) + static_cast<double>(IX2) * RM2) * RM1;

        ISEED = IX1;
        return result;
    }

private:
    std::vector<double> RDM;
    double RM1, RM2;
    int IA1, IC1, M1, IX1;
    int IA2, IC2, M2, IX2;
    int IA3, IC3, M3, IX3;
};

void testRandomNumberGenerator() {
    RandomNumberGenerator rng;
    int seed = 123456; // Example seed

    rng.DRNSET(seed);
    
    // Note: Fill in the expectedValues vector based on the output from the Fortran code for an accurate test
    std::vector<double> expectedValues = {/* Expected random values from Fortran output */};

    bool testPassed = true;
    for (size_t i = 0; i < expectedValues.size(); ++i) {
        double generated = rng.DRN(seed);
        if (std::fabs(generated - expectedValues[i]) > 1e-6) { // Allowing some floating point error
            std::cout << "Test failed at " << i + 1 << " Expected: " << expectedValues[i] << ", Got: " << generated << std::endl;
            testPassed = false;
        }
    }

    if (testPassed) {
        std::cout << "All tests passed!" << std::endl;
    }
}

int main() {
    testRandomNumberGenerator();
    return 0;
}

File: data/readable_format/fortran/file673.f90
PROGRAM TestRandomNumberGenerator
  IMPLICIT NONE
  INTEGER :: i, seed
  REAL*8 :: randomValue

  seed = 123456
  CALL DRNSET(seed)

  DO i = 1, 10
     randomValue = DRN(seed)
     PRINT *, 'Random number', i, ':', randomValue
  END DO

CONTAINS

  SUBROUTINE DRNSET( ISEED )
    IMPLICIT NONE
    INTEGER, INTENT(IN) :: ISEED
    INTEGER :: J
    REAL*8 :: RDM(31), RM1, RM2
    INTEGER :: IA1, IC1, M1, IX1
    INTEGER :: IA2, IC2, M2, IX2
    INTEGER :: IA3, IC3, M3, IX3

    COMMON /RANDM/ RDM, RM1, RM2, IA1, IC1, M1, IX1, IA2, IC2, M2, IX2, IA3, IC3, M3, IX3

    IA1 = 1279
    IC1 = 351762
    M1 = 1664557
    IA2 = 2011
    IC2 = 221592
    M2 = 1048583
    IA3 = 15091
    IC3 = 6171
    M3 = 29201

    IX1 = MOD( ISEED, M1 )
    IX1 = MOD( IA1*IX1+IC1, M1 )
    IX2 = MOD( IX1, M2 )
    IX1 = MOD( IA1*IX1+IC1, M1 )
    IX3 = MOD( IX1, M3 )
    RM1 = 1.0D0 / DBLE(M1)
    RM2 = 1.0D0 / DBLE(M2)
    DO J = 1, 31
       IX1 = MOD( IA1*IX1+IC1, M1 )
       IX2 = MOD( IA2*IX2+IC2, M2 )
       RDM(J) = ( DBLE(IX1) + DBLE(IX2)*RM2 )*RM1
    END DO
  END SUBROUTINE DRNSET

  REAL*8 FUNCTION DRN(ISEED)
    IMPLICIT NONE
    INTEGER, INTENT(INOUT) :: ISEED
    INTEGER :: J
    REAL*8 :: RDM(31), RM1, RM2
    INTEGER :: IA1, IC1, M1, IX1
    INTEGER :: IA2, IC2, M2, IX2
    INTEGER :: IA3, IC3, M3, IX3

    COMMON /RANDM/ RDM, RM1, RM2, IA1, IC1, M1, IX1, IA2, IC2, M2, IX2, IA3, IC3, M3, IX3

    IX1 = MOD( IA1*IX1+IC1, M1 )
    IX2 = MOD( IA2*IX2+IC2, M2 )
    IX3 = MOD( IA3*IX3+IC3, M3 )
    J = 1 + INT((31*IX3)/M3)
    DRN = RDM(J)
    RDM(J) = ( DBLE(IX1) + DBLE(IX2)*RM2 )*RM1

    ISEED = IX1
  END FUNCTION DRN

END PROGRAM TestRandomNumberGenerator